tensorflow/compiler/xla/service/gpu/cudnn_fused_conv_rewriter.cc

```diff
@@ -277,5 +293,223 @@ StatusOr<bool> CudnnFusedConvRewriter::Run(HloModule* module) {
   return changed;
 }
 
+// Describes a matched pattern:
+// convert_or_clamp(get_tuple_element(custom_call(x,w, ...)));
+// where the custom_call targets CuDNN convolution (either pure convolution or
+// fused convolution).
+struct ConvWithConvertOrClamp {
+  HloInstruction* convert_or_clamp;
+  HloInstruction* gte;
+  HloCustomCallInstruction* conv;
+};
+
+absl::optional<ConvWithConvertOrClamp> FindConvWithClampAndConvertToInt8(
+    HloInstruction* instr) {
+  using match::Broadcast;
+  using match::Clamp;
+  using match::Convert;
+  using match::GetTupleElement;
+  using match::Op;
+
+  // The pattern we want to match:
+  //   convert<int8>(clamp(broadcast(-128), (get_tuple_element(custom_call(int8_x,
+  //   int8_w, ...)), broadcast(127));
+  //
+
+  HloInstruction* gte = nullptr;
+  HloInstruction* conv_instr = nullptr;
+  auto lower_pattern = Broadcast(match::ConstantScalar(-128));
+  auto upper_pattern = Broadcast(match::ConstantScalar(127));
+  auto pattern =
+      Convert(Clamp(lower_pattern,
+            GetTupleElement(
+                &gte, Op(&conv_instr).WithOpcode(HloOpcode::kCustomCall), 0),
+            upper_pattern));
+
+  if (Match(instr, pattern)) {
+    if (conv_instr->operand(0)->shape().element_type() == xla::S8 &&
+        instr->shape().element_type() == xla::S8) {
+      HloCustomCallInstruction* conv =
+          CastOrNull<HloCustomCallInstruction>(conv_instr);
+      return ConvWithConvertOrClamp{instr, gte, conv};
+    }
+  }
+  return absl::nullopt;
+}
+
+StatusOr<bool> TryRewriteForConvertOrClampImpl(ConvWithConvertOrClamp match) {
```

Comment:
It appears that this function can not fail.
Should it be just `RewriteForConvertOrClampImpl`?

Reply:
I will rename the function name and return value for this and relevant functions.
Tests are added in the following PR (https://github.com/tensorflow/tensorflow/pull/30783/), as they need more changes there.  Please comment about any missing tests.

Reply:
Function names and return values have been updated.